geo/src/main/java/org/apache/carbondata/geo/scan/expression/PolygonListExpression.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.carbondata.core.scan.expression.conditional.ConditionalExpression;
+import org.apache.carbondata.core.scan.filter.executer.FilterExecutor;
+import org.apache.carbondata.core.scan.filter.intf.ExpressionType;
+import org.apache.carbondata.core.scan.filter.intf.RowIntf;
+import org.apache.carbondata.core.scan.filter.resolver.FilterResolverIntf;
+import org.apache.carbondata.core.scan.filter.resolver.RowLevelFilterResolverImpl;
+import org.apache.carbondata.core.util.CustomIndex;
+import org.apache.carbondata.geo.GeoConstants;
+import org.apache.carbondata.geo.GeoHashIndex;
+import org.apache.carbondata.geo.GeoHashUtils;
+import org.apache.carbondata.geo.scan.filter.executor.PolygonFilterExecutorImpl;
+
+/**
+ * InPolygonList expression processor. It inputs the InPolygonList string to the Geo
+ * implementation's query method, gets a list of range of IDs from each polygon and
+ * calculates the and/or/diff range list to filter as an output. And then, build
+ * InExpression with list of all the IDs present in those list of ranges.
+ */
+@InterfaceAudience.Internal
+public class PolygonListExpression extends UnknownExpression implements ConditionalExpression {
+
+  private String polygonListString;
+
+  private String opType;
+
+  private GeoHashIndex instance;
+
+  private List<Long[]> ranges = new ArrayList<Long[]>();
+
+  private ColumnExpression column;
+
+  private static final ExpressionResult trueExpRes =
+      new ExpressionResult(DataTypes.BOOLEAN, true);
+
+  private static final ExpressionResult falseExpRes =
+      new ExpressionResult(DataTypes.BOOLEAN, false);
+
+  public PolygonListExpression(String polygonListString, String opType, String columnName,
+      CustomIndex indexInstance) {
+    this.polygonListString = polygonListString;
+    this.opType = opType;
+    this.instance = (GeoHashIndex)indexInstance;
+    this.column = new ColumnExpression(columnName, DataTypes.LONG);
+  }
+
+  private void processExpression() {
+    try {
+      // 1. parse the polygon list string
+      List<String> polygons = new ArrayList<>();
+      Pattern pattern = Pattern.compile(GeoConstants.POLYGON_REG_EXPRESSION);
+      Matcher matcher = pattern.matcher(polygonListString);
+      while (matcher.find()) {
+        String matchedStr = matcher.group();
+        polygons.add(matchedStr);
+      }
+      if (polygons.size() < 2) {
+        throw new RuntimeException("polygon list need at least 2 polygons, really has " +
+            polygons.size());
+      }
+      // 2. get the range list of each polygon
+      List<Long[]> processedRangeList = instance.query(polygons.get(0));
+      for (int i = 1; i < polygons.size(); i++) {
+        List<Long[]> tempRangeList = instance.query(polygons.get(i));
+        processedRangeList = GeoHashUtils.processRangeList(
+            processedRangeList, tempRangeList, opType);
+      }
+      ranges = processedRangeList;
+      GeoHashUtils.validateRangeList(ranges);
```

Comment:
Intention of validateRangeList() to check if each element of List<Long[]> is only array of 2 elements(i.e., has only min and max value of range) was because `instance` class can be implemented by any user class adhering to `CustomIndex` interface. To make sure, `query()` returns list of array of only 2 elements, have this validation.
I think, we need to call `validateRangeList()` right after each call to `instance.query()`. Not required at the end after merge.

Reply:
Done. have checked all of `instance.query`